numpy/tests/test_public_api.py

```diff
@@ -746,3 +747,69 @@ def test___module__attribute():
 
     if incorrect_entries:
         assert len(incorrect_entries) == 0, incorrect_entries
+
+
+def _check___qualname__(obj, module_name: str) -> bool:
+    qualname = obj.__qualname__
+    name = obj.__name__
+    assert name == qualname.split(".")[-1]
+
+    if qualname == "show":
+        qualname = "show_config"
+    # Skip trimcoef from numpy.polynomial as it is duplicated by design.
+    if name == "trimcoef" and module_name.startswith("numpy.polynomial"):
+        return True
```

Comment:
Both of these look strange to me.  Can we please fix them in code and not in test?  

If this test fails the identity check (except for bound methods), then that is much worse than the `__module__` not pointing to the preferred public API exposure point, IMO.
For one, this is broken:  `pickle.loads(pickle.dumps(np.show_runtime))`  (maybe it always was).

The `trimcoef`, it seems to me like the identity check/pickling should work, so I am not sure why you need the skip?

Reply:
@seberg So `show` and `trimcoef` are two exceptions that IMO should be handled in a special way. It comes from the fact that they are imported with a different name than the actual `__name__` and `__qualname__` and I don't think it has anything in common with `__module__` here:

1. For `trimcoef`, the issue is that it's a function that is defined in `numpy.polynomial.polyutils` with a name `trimcoef` with `__name__` and `__qualname__` of `trimcoef` (as it's just a function). Then this function is re-exported to a few modules with a different name, namely `np.polynomial.legendre` as `legtrim`, `np.polynomial.hermite` as `hermtrim`, etc. For each of these exports the `__(qual)name__` is still `trimcoef` so our test fails here as `functools.reduce(getattr, qualname.split("."), module)` can't find it. Ideally `trimcoef` should be available in one place and shouldn't be re-exported with a different name in 5 different modules but I think we decided to skip this refactor for 2.0 (and we do exactly the same special path in `test_functions_single_location` [link](https://github.com/numpy/numpy/blob/29306a293d11fa9db0e22aa390fc0e29e300c61a/numpy/tests/test_public_api.py#L644)). I would prefer to have `trimcoef` in one place though! But for now there's nothing to fix in codebase without braking API and we need to handle it in a test.
2. `show`/`show_config` is somewhat similar (not to be confused with `np.show_runtime`, this one works correctly in our test). Function `show` is exported in `numpy` namespace with `show_config` name (`from numpy.__config__ import show as show_config`) so when we access `__(qual)name__`, we can't find it in the `numpy` namespace. To fix it, maybe we can rename `show` in `__config__` to `show_config`? But then I'm surprised we didn't do in in NEP 52 for 2.0 release.

WDYT?

Reply:
`__module__` changed for `np.show_config` and it indeed breaks pickling. I can either:
1. Make `__module__` point to `numpy.__config__` again.
2. Rename `def show` to `def show_config`.
3. Additionally export `show` as `show` to the main namespace.

Reply:
For `trimcoef` you can then write the test to use `obj.__module__` rather than the module you found it in.  The module where it appears doesn't actually matter since that is the job of `test_functions_single_location` where the special case is explained.

For `show_config` please suggest something can you please do what you think is best and I'll review it?

Reply:
Thank you for guidance!

I updated `trimcoef` case in the test so it uses `obj.__module__` now.

For `show_config` I decided to just add another `def show_config` function with correct `__module__`, `__(qual)name__`, and `__doc__`, which just calls `np.__config__.show` underneath. This way we are backward compatible and don't break `np.__config__.show()` calls, as there are [plenty of them](https://github.com/search?q=%22np.__config__.show%22&type=code), and we can have correct `np.show_config` function without weird hacks. WDYT?


P.S. Added misssing typing stubs and typing tests updates for UFuncs.